staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1628 @@
+        VIRTIO_GPU_MAX_SCANOUTS,
+    },
+    virtio_gpu::{RutabagaVirtioGpu, VirtioGpu, VirtioGpuRing},
+    GpuConfig, GpuMode,
+};
+use log::{debug, error, trace, warn};
+use rutabaga_gfx::{
+    ResourceCreate3D, RutabagaFence, Transfer3D, RUTABAGA_PIPE_BIND_RENDER_TARGET,
+    RUTABAGA_PIPE_TEXTURE_2D,
+};
+use std::{
+    cell::RefCell,
+    io::ErrorKind,
+    io::{self, Result as IoResult},
+    os::fd::AsRawFd,
+    sync::{self, Arc, Mutex},
+};
+use thiserror::Error as ThisError;
+use vhost::vhost_user::{
+    gpu_message::{VhostUserGpuCursorPos, VhostUserGpuEdidRequest},
+    message::{VhostUserProtocolFeatures, VhostUserVirtioFeatures},
+    GpuBackend,
+};
+use vhost_user_backend::{VhostUserBackend, VringEpollHandler, VringRwLock, VringT};
+use virtio_bindings::{
+    bindings::{
+        virtio_config::{VIRTIO_F_NOTIFY_ON_EMPTY, VIRTIO_F_RING_RESET, VIRTIO_F_VERSION_1},
+        virtio_ring::{VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC},
+    },
+    virtio_gpu::{
+        VIRTIO_GPU_F_CONTEXT_INIT, VIRTIO_GPU_F_EDID, VIRTIO_GPU_F_RESOURCE_BLOB,
+        VIRTIO_GPU_F_VIRGL,
+    },
+};
+use virtio_queue::{QueueOwnedT, Reader, Writer};
+use vm_memory::{ByteValued, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap, Le32};
+use vmm_sys_util::{
+    epoll::EventSet,
+    eventfd::{EventFd, EFD_NONBLOCK},
+};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+pub enum Error {
+    #[error("Failed to handle event, didn't match EPOLLIN")]
+    HandleEventNotEpollIn,
+    #[error("Failed to handle unknown event")]
+    HandleEventUnknown,
+    #[error("Descriptor read failed")]
+    DescriptorReadFailed,
+    #[error("Descriptor write failed")]
+    DescriptorWriteFailed,
+    #[error("Invalid command type {0}")]
+    InvalidCommandType(u32),
+    #[error("Failed to send used queue notification: {0}")]
+    NotificationFailed(io::Error),
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+    #[error("Failed to create an iterator over a descriptor chain: {0}")]
+    CreateIteratorDescChain(virtio_queue::Error),
+    #[error("Failed to create descriptor chain Reader: {0}")]
+    CreateReader(virtio_queue::Error),
+    #[error("Failed to create descriptor chain Writer: {0}")]
+    CreateWriter(virtio_queue::Error),
+    #[error("Failed to decode gpu command: {0}")]
+    GpuCommandDecode(GpuCommandDecodeError),
+    #[error("Failed to encode gpu response: {0}")]
+    GpuResponseEncode(GpuResponseEncodeError),
+    #[error("Failed add used chain to queue: {0}")]
+    QueueAddUsed(virtio_queue::Error),
+    #[error("Epoll handler not available: {0}")]
+    EpollHandler(String),
+    #[error("Failed register epoll listener: {0}")]
+    RegisterEpollListener(io::Error),
+}
+
+impl From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
```

Comment:
This looks weird to me, do you need it? I don't expect this crate to ever return an io::Error when you already have a Error type that covers an IO kind.

Reply:
yes, the methods called in `fn handle_event` need this. Also other crates  Implement this `From<Error> for std::io::Error`